File: src/main/java/juuxel/accesswidener/idea/psi/impl/AwPsiImplUtil.java
package juuxel.accesswidener.idea.psi.impl;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.tree.IElementType;
import juuxel.accesswidener.idea.AccessType;
import juuxel.accesswidener.idea.psi.AwClassDefinition;
import juuxel.accesswidener.idea.psi.AwDefinition;
import juuxel.accesswidener.idea.psi.AwHeader;
import juuxel.accesswidener.idea.psi.AwMemberDefinition;
import juuxel.accesswidener.idea.psi.AwMethodDescriptor;
import juuxel.accesswidener.idea.psi.AwTypeDescriptor;
import juuxel.accesswidener.idea.psi.AwTypes;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.function.Function;

public final class AwPsiImplUtil {
    @Nullable
    private static <T> T ofChild(PsiElement elt, IElementType type, Function<ASTNode, T> fn) {
        ASTNode node = elt.getNode().findChildByType(type);
        return node != null ? fn.apply(node) : null;
    }

    @Nullable
    private static String textOfChild(PsiElement elt, IElementType type) {
        return ofChild(elt, type, ASTNode::getText);
    }

    @Nullable
    public static AccessType getAccessType(AwDefinition definition) {
        return ofChild(definition, AwTypes.ACCESS, AwPsiImplUtil::getAccessType);
    }

    @Nullable
    public static AccessType getAccessType(ASTNode access) {
        switch (access.getText()) {
            case "accessible":
                return AccessType.ACCESSIBLE;
            case "extendable":
                return AccessType.EXTENDABLE;
            case "mutable":
                return AccessType.MUTABLE;
            default:
                return null;
        }
    }

    @Nullable
    public static String getOwner(AwMemberDefinition definition) {
        return textOfChild(definition, AwTypes.CLASS_NAME);
    }

    @Nullable
    public static PsiElement getMemberIdentifier(AwMemberDefinition definition) {
        return ofChild(definition, AwTypes.MEMBER_NAME, ASTNode::getPsi);
    }

    @Nullable
    public static String getName(AwDefinition definition) {
        if (definition instanceof AwClassDefinition) {
            return textOfChild(definition, AwTypes.CLASS_NAME);
        } else {
            return textOfChild(definition, AwTypes.MEMBER_NAME);
        }
    }

    @Nullable
    public static Integer getVersionNumber(AwHeader header) {
        return ofChild(header, AwTypes.VERSION, node -> Integer.parseInt(node.getText().substring(1)));
    }

    @Nullable
    public static String getNamespace(AwHeader header) {
        return textOfChild(header, AwTypes.NAMESPACE);
    }

    @NotNull
    public static String getDescriptorString(AwTypeDescriptor descriptor) {
        return descriptor.getText();
    }

    public static boolean isPrimitive(AwTypeDescriptor descriptor) {
        return descriptor.getNode().findChildByType(AwTypes.PRIMITIVE_DESCRIPTOR) != null;
    }

    public static boolean isLiteral(AwTypeDescriptor descriptor) {
        return !isPrimitive(descriptor);
    }

    @Nullable
    public static String getClassName(AwTypeDescriptor descriptor) {
        return textOfChild(descriptor, AwTypes.CLASS_NAME);
    }

    @NotNull
    public static AwTypeDescriptor getReturnType(AwMethodDescriptor descriptor) {
        List<AwTypeDescriptor> descriptors = descriptor.getTypeDescriptorList();
        return descriptors.get(descriptors.size() - 1);
    }

    @NotNull
    public static List<AwTypeDescriptor> getParameters(AwMethodDescriptor descriptor) {
        List<AwTypeDescriptor> descriptors = descriptor.getTypeDescriptorList();

        if (descriptors.size() == 1) return Collections.emptyList();
        return descriptors.subList(0, descriptors.size() - 1);
    }
}

File: src/main/kotlin/juuxel/accesswidener/idea/psi/impl/AwPsiImplUtil.kt
package juuxel.accesswidener.idea.psi.impl

import com.intellij.icons.AllIcons
import com.intellij.lang.ASTNode
import com.intellij.navigation.ItemPresentation
import com.intellij.navigation.ItemPresentationProviders
import com.intellij.openapi.util.Iconable
import com.intellij.psi.JavaPsiFacade
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiType
import com.intellij.psi.tree.IElementType
import juuxel.accesswidener.idea.AccessType
import juuxel.accesswidener.idea.psi.AwClassDefinition
import juuxel.accesswidener.idea.psi.AwDefinition
import juuxel.accesswidener.idea.psi.AwFieldDefinition
import juuxel.accesswidener.idea.psi.AwHeader
import juuxel.accesswidener.idea.psi.AwMemberDefinition
import juuxel.accesswidener.idea.psi.AwMethodDefinition
import juuxel.accesswidener.idea.psi.AwMethodDescriptor
import juuxel.accesswidener.idea.psi.AwTypeDescriptor
import juuxel.accesswidener.idea.psi.AwTypes
import juuxel.accesswidener.idea.util.Types
import javax.swing.Icon

object AwPsiImplUtil {
    private inline fun <T> ofChild(elt: PsiElement, type: IElementType, fn: (ASTNode) -> T): T? =
        elt.node.findChildByType(type)?.let(fn)

    private fun textOfChild(elt: PsiElement, type: IElementType): String? =
        ofChild(elt, type, ASTNode::getText)

    fun getAccessType(access: ASTNode): AccessType? =
        when (access.text) {
            "accessible" -> AccessType.ACCESSIBLE
            "extendable" -> AccessType.EXTENDABLE
            "mutable" -> AccessType.MUTABLE
            else -> null
        }

    // AwDefinition

    @JvmStatic
    fun getAccessType(definition: AwDefinition): AccessType? =
        ofChild(definition, AwTypes.ACCESS, ::getAccessType)

    @JvmStatic
    fun getName(definition: AwDefinition): String? =
        when (definition) {
            is AwClassDefinition -> textOfChild(definition, AwTypes.CLASS_NAME)
            else -> textOfChild(definition, AwTypes.MEMBER_NAME)
        }

    @JvmStatic
    fun getIcon(definition: AwDefinition, @Iconable.IconFlags flags: Int): Icon =
        getIcon(definition)

    @JvmStatic
    fun getIcon(definition: AwDefinition): Icon =
        when (definition) {
            is AwClassDefinition -> AllIcons.Nodes.Class
            is AwMethodDefinition -> AllIcons.Nodes.Method
            is AwFieldDefinition -> AllIcons.Nodes.Field
            else -> throw IllegalArgumentException("Unknown definition: $definition")
        }

    @JvmStatic
    fun getPresentation(definition: AwDefinition): ItemPresentation? =
        ItemPresentationProviders.getItemPresentation(definition)

    // AwMemberDefinition
    // All methods are guaranteed to be nonnull because members are pinned on their descriptors

    @JvmStatic
    fun getOwner(definition: AwMemberDefinition): String =
        textOfChild(definition, AwTypes.CLASS_NAME)!!

    @JvmStatic
    fun getMemberIdentifier(definition: AwMemberDefinition): PsiElement =
        ofChild(definition, AwTypes.MEMBER_NAME, ASTNode::getPsi)!!

    @JvmStatic
    fun getName(definition: AwMemberDefinition): String =
        textOfChild(definition, AwTypes.MEMBER_NAME)!!

    @JvmStatic
    fun getDescriptor(definition: AwMemberDefinition): PsiElement =
        when (definition) {
            is AwMethodDefinition -> definition.methodDescriptor
            is AwFieldDefinition -> definition.typeDescriptor
            else -> throw IllegalArgumentException("Unknown AW member: $definition")
        }

    // AwHeader

    @JvmStatic
    fun getVersionNumber(header: AwHeader): Int? =
        ofChild(header, AwTypes.VERSION) { it.text.substring(1).toInt() }

    @JvmStatic
    fun getNamespace(header: AwHeader): String? =
        textOfChild(header, AwTypes.NAMESPACE)

    // AwTypeDescriptor

    @JvmStatic
    fun getDescriptorString(descriptor: AwTypeDescriptor): String =
        descriptor.text

    @JvmStatic
    fun isPrimitive(descriptor: AwTypeDescriptor): Boolean =
        descriptor.node.findChildByType(AwTypes.PRIMITIVE_DESCRIPTOR) != null

    @JvmStatic
    fun isLiteral(descriptor: AwTypeDescriptor): Boolean =
        !isPrimitive(descriptor)

    @JvmStatic
    fun getClassName(descriptor: AwTypeDescriptor): String? =
        textOfChild(descriptor, AwTypes.CLASS_NAME)?.let(Types::toJavaName)

    @JvmStatic
    fun toPsiType(descriptor: AwTypeDescriptor): PsiType? =
        if (isPrimitive(descriptor)) {
            when (val desc = descriptor.descriptorString) {
                "B" -> PsiType.BYTE
                "S" -> PsiType.SHORT
                "I" -> PsiType.INT
                "J" -> PsiType.LONG
                "F" -> PsiType.FLOAT
                "D" -> PsiType.DOUBLE
                "Z" -> PsiType.BOOLEAN
                "C" -> PsiType.CHAR
                "V" -> PsiType.VOID
                else -> error("Unknown primitive descriptor: $desc")
            }
        } else {
            getClassName(descriptor)?.let { name ->
                JavaPsiFacade.getElementFactory(descriptor.project)
                    .createTypeByFQClassName(name)
            }
        }

    // AwMethodDescriptor

    // TODO: Investigate these since they are naive wrt erroring parsings

    @JvmStatic
    fun getReturnType(descriptor: AwMethodDescriptor): AwTypeDescriptor =
        descriptor.typeDescriptorList.last()

    @JvmStatic
    fun getParameters(descriptor: AwMethodDescriptor): List<AwTypeDescriptor> =
        descriptor.typeDescriptorList.dropLast(1)
}
